Inter filter

ernfrid opened the pull request:
Here it is in its current state. If you have any suggestions on ways to make the code better, please let me know so I can fix it up as best I can.

ernfrid commented:
Also - I'm planning on squashing before actually opening a pull request so apologies for the abundance of commits.

brentp commented:
no problem I can squash through the github UI. Running this on GiaB now.

ernfrid commented on lumpy/depthfilter.go:
```diff
@@ -369,6 +370,33 @@ func cp(dst, src string) error {
+
+func localOutOfWindow(i, k inter) bool {
+	return (k.read1_tid > i.read1_tid || (k.read1_tid == i.read1_tid && k.read1_pos > i.read1_pos + 300))
+}
```
That's a good point on the window size. I've only tested the 300 bp option and results looked good. The first idea I'd had was that 2 * read length should be sufficient, but perhaps it'd be more appropriate to be something like mean/median insert size ± 3 * mad/sd.

ernfrid replied:
For my test sample, the standard deviation (from the lumpy command line) was 78bp.

brentp replied:
yeah, it should be on the insert-size, not on the read-length. In looking through IGV at stuff that changed, I did see a true read that got removed by this as it was just over 300 bp. I think the right side can be much larger, e.g. 10KB since that's unlikely that a bad read will land within 2KB in the all vs all genome space.

brentp replied:
I am trying on giab with 10kb now.

ernfrid replied:
Sounds good. Alternatively, seems like we could pass in the stats from the filter struct and calculate a dynamic filter.

brentp replied:
here are the numbers from master, your PR, your PR with the const changed to 1000:
brentp@lorax:~/go/src/github.com/brentp/smoove$ samtools view -c hg002/hg002.disc.bam
1434810
brentp@lorax:~/go/src/github.com/brentp/smoove$ samtools view -c hg002-ernfrid/hg002.disc.bam
331256
brentp@lorax:~/go/src/github.com/brentp/smoove$ samtools view -c hg002-ernfrid10k/hg002.disc.bam
333506
```

given we get only a few extra reads with 10KB, I think that's a safer default.

brentp replied:
wow. that's suprising. well, I guess using the covstats.Stats.TemplateMean + 4 * covstats.Stats.TemplateSD should be somewhere in the middle.

ernfrid commented on lumpy/depthfilter.go:
```diff
@@ -388,6 +418,28 @@ func singletonfilter(fbam string, split bool, originalCount int) {
 				name = "A" + name[1:]
+			} else {
+				if interOrDistant(rec) {
+					cur := inter{read1_tid: rec.Ref.ID(), read2_tid: rec.MateRef.ID(), read1_pos: rec.Pos, read2_pos: rec.MatePos, qname: name}
```
Added...

brentp commented:
if you're up to it, you could convert the tids and read_pos's too uint32. if we have 70 million of those in memory, it will make a difference. if not, I can do this after this PR is merged.

ernfrid commented:
I'll take a stab at it. Futzing with the windowing code at the moment.

ernfrid commented:
So there are some problems with the current implementation. I added some tests that demonstrate the issue. 

My intuition is that these probably get worse as the window size gets bigger and that is contributing to why I'm not seeing as a big a decrease as you'd expect with a 10kb window.

I also think I added in uint32 support though I haven't tackled the integration with library insert size as I think I need a new filter implementation first.

ernfrid commented:
I still have some refactoring to do and need to update the tests some more, but can you switch the window to 10kb and try again on GiaB? I think this version addresses the issue and I'm getting decent performance with 10kb.

brentp commented:
let me know when you're reading for me to merge. I can squash in the github UI.

ernfrid commented:
This is squashed and ready as far as I'm concerned.
